packages/providers/src/base.provider.ts

```diff
@@ -0,0 +1,106 @@
+import {
+  camelCase,
+  capitalCase,
+  constantCase,
+  kebabCase,
+  pascalCase,
+  snakeCase,
+} from './utils/change-case';
+import { IOptions } from './utils/change-case/functions';
+import { deepmerge } from './utils/deepmerge.utils';
+import { WithPassthrough } from './utils/types';
+
+export enum CasingEnum {
+  CAMEL_CASE = 'camelCase',
+  PASCAL_CASE = 'PascalCase',
+  CAPITAL_CASE = 'CapitalCase',
+  SNAKE_CASE = 'snake_case',
+  KEBAB_CASE = 'kebabcase',
+  CONSTANT_CASE = 'CONSTANT_CASE',
+}
+
+type TransformOutput<T> = {
+  body: T;
+  headers: Record<string, string>;
+  query: Record<string, string>;
+};
+
+export abstract class BaseProvider {
+  protected casing: CasingEnum | undefined;
+
+  protected keyCaseObject: Record<string, string> = {};
+
+  protected transform<
+    Output = Record<string, unknown>,
```

Comment:
I usually prefer to prefix generic parameters with `T_` to make it very clear that they are generic params, otherwise they can be confused with `TrainCased` class names.

```suggestion
    T_Output = Record<string, unknown>,
```